doc/python/hardware.rst

```diff
@@ -152,9 +152,8 @@ The Python wrapper provides an API for:
 
         .. attribute:: threads
 
-            The number of threads available locally for execution. Must be set to 1 at minimum. 1 by default.
-            Passing ``"avail_threads"`` (as string) will query and use the maximum number of threads the system makes available.
-
+            The number of threads available locally for execution. Defaults to the maximum number of threads the system makes available if gpu_id and mpi are not set, else defaults to 1.
```

Comment:
Considering the rest of the change set, how about digging in a bit deeper and come up with sensible defaults for _all_ cases? It's not a requirement to merge, but if we are going to put it in eventually, why not in 0.8?
I don't know whether arbenv has the proper calls, but
- MPI: figure out how many threads a task is allowed to use (using hwloc maybe) and set that when handed 'avail_threads'. Other than that there's OS-specific calls (which hwloc uses), we would need to consider MacOS and Linux.
- GPU: can just use 'avail_threads', it speeds up construction a bit. Independent of the change above. If 'avail_threads' receives super-charged capabilities wrt MPI, it'll benefit, if not, then not.
- MPI+GPU: do whatever MPI does. Likely we are using one task per GPU.

Reply:
My initial idea was to only touch the default constructor, to provide the best experience for novice users.

If you'd like to touch more than the default constructor: we could initialize `threads` to `thread_concurrency`. It's straightforward to understand for the user, but in particular when MPI is used, may result in surprising behavior.

Reply:
MPI config is it's own world of expertise, and I wager normal users stick to 1 thread per rank (I have never seen or done differently, unless it was you trying to squeeze out every last drop).

So, my pref is then to default `threads` to `thread_concurrency` unless `mpi` is set, in which case it stays `1`. If cleverer people than I come up with a thought out proposal for MPI then I'm happy to consider it.

Reply:
As I said, I'd be fine with merging this part as is, but consider the extra effort for making it extra-nice for the user interesting. I think the common patterns are
1. No MPI, 1 thread/(logical)core 
  - with GPU 
  - without GPU
3. MPI, still 1 thread per (logical)core but now each task gets handed `1/n`th. N is the number of tasks according to
  - 1 task per socket
  - 1 task GPU

So, the approach is to compute 
- the (logical)core count: `arbenv::cores(logical=true|false)`
- the number of tasks: handed to you via `Comm_size`, defaults to 1

and `context` or `thread_concurrency` could be instructed whether to distribute threads over 
tasks or not.

Reply:
See above's note on MPI.

Reply:
As I said, it'd been nice to have rather than critical. I am personally fine with typing out these, but as we were sanding off some rougher edges it offered an easy win.

Reply:
I'd want to test this out, but for this week I don't think I'll have the time. Making the release will almost certainly eat up the rest of my hours, and I prefer not to postpone it further. Maybe a small discussion in the team meet about removing proc_alloc and have these new constructors?

Reply:
As said, optional, and only if `proc_alloc` blocks you from achieving more.